[IMP] l10n_sa, l10n_sa_edi: Improved Report Spacing

alny-odoo opened the pull request:
Description of the issue/feature this PR addresses:
Task ID: 4342267

Current behavior before PR:
-Error When printing draft invoice due to unset invoice_date.

-Report spacing between header & body in second page overlaps due to paperformat spacing.
Due to the long address format of saudi arabia, the default paperformat header spacing and margin top are not enough to show the full address without overlapping the second and onward pages.

-English labels on the exchange rate table in l10n_sa_edi are translated and not shown together with the arabic labels.

-Remove unnecessary spans in xpaths

Desired behavior after PR is merged:
-Fix printing draft invoice with invoice_date as today's date

-Fix report spacing header by adding a new A4 paperformat for l10n_sa which is set as the saudi arabia companies default paper format.

-Fix exchange rate table to show both arabic and english labels and remove unnecessary translations (no longer used)

-Remove unnecessary spans in xpath.



---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/190172.png)](https://mergebot.odoo.com/odoo/odoo/pull/190172)

proose commented on addons/l10n_sa_edi/views/report_invoice.xml:
```diff
@@ -7,21 +7,22 @@
-                        <td style="width: 25%"><strong>Subtotal (SAR)</strong></td>
-                        <td style="width: 25%"><strong>VAT Amount (SAR)</strong></td>
-                        <td style="width: 25%"><strong>Total (SAR)</strong></td>
+                        <td style="width: 25%"><strong>سعر الصرف<br/>Exchange Rate</strong></td>
```
Better use the (bootstrap) class w-25 which should do the same

proose commented on addons/l10n_sa_edi/views/report_invoice.xml:
```diff
@@ -7,21 +7,22 @@
             <xpath expr="//div[hasclass('clearfix')]" position="after">
                 <table t-if="o.company_id.country_id.code == 'SA' and o.currency_id != o.company_id.currency_id"
-                     id="sar_amounts" t-att-style="'ltr' if lang != 'ar_001' else 'rtl'" class="row clearfix ms-auto my-3 table table-sm table-borderless">
+                    id="sar_amounts" style="page-break-inside: avoid;" class="clearfix mx-auto my-3 table-sm table-borderless">
```
Does bootstrap have a class for that ?

alny-odoo replied:
I tried finding a class for it but couldn't find one.

proose commented on addons/l10n_sa/views/report_invoice.xml:
```diff
@@ -102,11 +107,18 @@
         <xpath expr="//div[@name='invoice_date']//span[@t-field='o.invoice_date']" position="attributes">
             <attribute name="t-if">not o.l10n_sa_confirmation_datetime</attribute>
         </xpath>
+        <xpath expr="//p[@name='payment_term']//div" position="attributes">
```
Why do you add this ? From what I see, this is always only one line on the invoice, did you find some exceptions ?

proose replied:
Why do you add this ? From what I see, this is always only one line on the pdf (so it already shouldn't break IMO), did you find some exceptions ?

proose reviewed (changes requested):
Hello, thank you for your work !
However I have some issues with it for now:
- the header is now too high IMO (too close to the top of the page)
- don't forget to add a message to your commit
- have you tried upgrading with your code ? Have a db with l10n_sa_edi, add your commit to the code and check that no issue arise (no issue in terminal and functionally). Then upgrade the module with and check again. I'm asking this because I'm not sure about the changes to translations and the new paperformat.

You might want to check the functional result with a PO to be sure the changes are what we want :)

alny-odoo commented:
HI @proose I have adjusted the margin top to give more space to the top of the page. also added a commit message.

regarding the new paperformat. I have placed the xml function inside a data element rather than a noupdate so that it triggers during an upgrade of the module. I have tested it after this change and it's being set on the SA companies.

proose reviewed (commented):
Thank you for the changes, I checked with OLMA who agreed with what he saw from the changes.
Another reviewer will come take a last look ;)

alny-odoo commented on addons/l10n_sa_edi/i18n/ar.po:
```diff
@@ -113,26 +113,6 @@ msgstr ""
 "<span class=\"o_form_label\">دمج الواجهة البرمجية لزاتكا</span>\n"
 "                            <span class=\"fa fa-lg fa-building-o\" title=\"Values set here are company-specific.\" aria-label=\"Values set here are company-specific.\" groups=\"base.group_multi_company\" role=\"img\"/>"
 
-#. module: l10n_sa_edi
```
Hi, if it's possible then I've removed it again, because it's no longer needed as I use both arabic and english in the edi labels now.

gawa-odoo commented on addons/l10n_sa_edi/views/report_invoice.xml:
```diff
@@ -7,21 +7,22 @@
-                        <td style="width: 25%"><strong>Subtotal (SAR)</strong></td>
-                        <td style="width: 25%"><strong>VAT Amount (SAR)</strong></td>
-                        <td style="width: 25%"><strong>Total (SAR)</strong></td>
+                        <td class="w-25"><strong>سعر الصرف<br/>Exchange Rate</strong></td>
+                        <td class="w-25"><strong>الإجمالي الفرعي بالريال السعودي<br/>Subtotal</strong></td>
+                        <td class="w-25"><strong>مبلغ ضريبة القيمة المضافة بالريال السعودي<br/>VAT Amount</strong></td>
+                        <td class="w-25"><strong>الإجمالي بالريال السعودي<br/>Total</strong></td>
```
Shouldn't these be in rtl?

alny-odoo replied:
hi @gawa-odoo I fixed this by splitting the arabic and english into different tds so we can apply the rtl to the arabic labels.

I've also adjusted the commit message to be more descriptive and include the task id.

Also brought back the translations.

however since I'm bringing back the translations I had to replace the <strong> english labels in l10n_sa_edi to be span with class fw-bold so the translations don't affect them anymore as the arabic and english versions are always shown

gawa-odoo replied:
@alny-odoo 

> however since I'm bringing back the translations I had to replace the \<strong\> english labels in l10n_sa_edi to be span with class fw-bold so the translations don't affect them anymore as the arabic and english versions are always shown

Fine by me.

But it seems to me that the text in arabic is still left aligned? 
![image](https://github.com/user-attachments/assets/0be71466-1afa-40b6-adc4-2db566f25de6)

alny-odoo replied:
Hi @gawa-odoo yes this was requested by Yatin.

gawa-odoo reviewed (commented):
Hello!
I'm not sure you should remove the translations in po/pot, for the people who don't update their modules or things like that. If I'm not mistaken, in general, we leave them.

And I would appreciate if you could improve a bit your commit message. It does not help to understand what you're doing (except the paperformat)
Your PR message specifies more what you're trying to fix than `minor fixes` for instance.
Also don't hesitate to put the task id in the commit message.

gawa-odoo commented:
@robodoo r+
